docs(node-agent): add proposal (#1)
dev-sys-do/orka: docs/proposals/node-agent/README.md

# Node Agent

## Requirements

### Scope

The node agent manages `orka` workloads on a specific node. There is one agent per node.

* A node is a physical or virtual machine in the cluster, that is managed by the `orka` control plane.
* A workload is the smallest unit of compute deployable in orka. It can be e.g. a container, a virtual machine, etc.

The Node Agent provides an API  for controlling workloads lifecycle on the node it manages, and a node monitoring interface:

- to control the workloads :
  - start
  - stop
  - restart
  - kill
- health of node
- metrics of node
- status of workloads
  - waiting
  - running
  - terminated
- keeping workloads running/alive
- cleaning up terminated workloads

#### Out of scope

- Networking configuration.
- Workload and node status storage. This is the responsibility of the `orka` controller.
- Workload rescheduling.

## High level architecture

### Interfaces

> See agent.proto

Reviewer: ```suggestion
See the [node agent gRPC specification](docs/proposals/node-agent/agent.proto)
```

# Node Agent

## Requirements

### Which service must the domain provide

> The node agent amanges workload management. A workload is the tiniest unit in orka. It can be a container, a virtual machine, etc.

Reviewer: ```suggestion
The node agent manages `orka` workloads on a specific node. There is one agent per node.

* A node is a physical or virtual machine in the cluster, that is managed by the `orka` control plane.
* A workload is the smallest unit of compute deployable in orka. It can be e.g. a container, a virtual machine, etc.
```

> ### What is in scope and out of scope
> 
> #### In scope

Reviewer: ```suggestion
### Scope
```

> The Node Agent provides an API :

Reviewer: ```suggestion
The Node Agent provides an API  for controlling workloads lifecycle on the node it manages, and a node monitoring interface:
```

Reviewer: Could you please address that comment?

- to control the workloads :
  - start
  - stop

>   - restart

Reviewer: You also want a `kill` operation on workloads.

Reviewer 2: Agreed, fixed in : 412e498be7f48ac42ff799628d03add4272d6742

- health of node
- metrics of nodes
- status of workloads
  - waiting
  - running

>   - terminated

Reviewer 3: This describes that `Status` should be called by the Scheduler. Instead, the docs (namely slide n°79) stipulates that:

> Agent sends nodes and instances updates.

Please review the Scheduler's `agent.proto` implementation so we can plan together.

Reviewer: That's correct. The proposed architecture has the agent sending nodes and instances updates to the scheduler.

> - to control the workloads :
>   - start
>   - stop
>   - restart
> - health of node
> - metrics of nodes
> - status of workloads
>   - waiting
>   - running
>   - terminated

Reviewer: ```suggestion
#### Workload Lifecycle
  - `start`
  - `stop`
  - `restart`
  
#### Node Monitoring
- Health
- Metrics

```

- keeping workloads running/alive

#### Out of scope

> - networking

Reviewer: It's not entirely out of scope, as it's expecting some hooks into the networking provider/interface. What's out of scope is the networking configuration.

Reviewer 2: fixed in : ed97b9fb54f7f5c45b381b1f47aea383707cb1f0

> - saving the node state in the database
> - saving the workload state in the database

Reviewer: ```suggestion
- Workload and node status storage. This is the responsibility of the `orka` controller.
```

Reviewer 2: fixed in : ed97b9fb54f7f5c45b381b1f47aea383707cb1f0

> - rescheduling workloads

Reviewer: Yes, that's a good point. A workload instance is never started intentionally by the agent.

> - cleaning up terminated workloads

Reviewer: This *is* in scope. When a workload is terminated, crashed or killed, the agent will have to clean the mess.

Reviewer 2: fixed in : 04f036979b3704583c912566e7106750b3b77934

# Node Agent

## Requirements

> ### Which service must the domain provide

Reviewer: I would merge that section under the `Scope` one. See below.

The node agent manages `orka` workloads on a specific node. There is one agent per node.

* A node is a physical or virtual machine in the cluster, that is managed by the `orka` control plane.
* A workload is the smallest unit of compute deployable in orka. It can be e.g. a container, a virtual machine, etc.

> ### Scope

Reviewer: ```suggestion
### Scope

The node agent manages `orka` workloads on a specific node. There is one agent per node.

* A node is a physical or virtual machine in the cluster, that is managed by the `orka` control plane.
* A workload is the smallest unit of compute deployable in orka. It can be e.g. a container, a virtual machine, etc.

```

Reviewer 4: Fixed in https://github.com/dev-sys-do/orka/pull/1/commits/4c226c00975c5232c7216a4d699092bc85f346d6

The Node Agent provides an API :

- to control the workloads :
  - start
  - stop
  - restart
  - kill
- health of node
- metrics of node
- status of workloads
  - waiting
  - running
  - terminated
- keeping workloads running/alive
- cleaning up terminated workloads

#### Out of scope

> - networking configuration
> - workload and node status storage. This is the responsibility of the `orka` controller.
> - rescheduling workloads

Reviewer: Nitty nits:

```suggestion
- Networking configuration.
- Workload and node status storage. This is the responsibility of the `orka` controller.
- Workload rescheduling.
```

Reviewer 4: Fixed in https://github.com/dev-sys-do/orka/pull/1/commits/4c226c00975c5232c7216a4d699092bc85f346d6

> ## Propose high level architecture

Reviewer: ```suggestion
## High level architecture
```

Reviewer 4: Fixed in https://github.com/dev-sys-do/orka/pull/1/commits/4c226c00975c5232c7216a4d699092bc85f346d6

> ## Propose API or commands

Reviewer: Please remove that section

Reviewer 4: Done in https://github.com/dev-sys-do/orka/pull/1/commits/4c226c00975c5232c7216a4d699092bc85f346d6

> ### What is your component interface ?

Reviewer: ```suggestion
### Interfaces
```

Reviewer 4: Fixed in https://github.com/dev-sys-do/orka/pull/1/commits/4c226c00975c5232c7216a4d699092bc85f346d6
